File: packages/compiler/src/check.ts
import ts from "typescript";
import type { Diagnostic } from "./diagnostics.ts";
import * as typescript from "./typescript.ts";

export type Request = {
	modules: Array<string>;
};

export type Response = {
	diagnostics: Array<Diagnostic>;
};

export let handle = (request: Request): Response => {
	let diagnostics = [];

	// Create a typescript program.
	let program = ts.createProgram({
		rootNames: request.modules.map(typescript.fileNameFromModule),
		options: typescript.compilerOptions,
		host: typescript.host,
	});

	// Collect the TypeScript diagnostics.
	diagnostics.push(
		...[
			...program.getConfigFileParsingDiagnostics(),
			...program.getOptionsDiagnostics(),
			...program.getGlobalDiagnostics(),
			...program.getDeclarationDiagnostics(),
			...program.getSyntacticDiagnostics(),
			...program.getSemanticDiagnostics(),
		].map(typescript.convertDiagnostic),
	);

	return { diagnostics };
};

File: packages/compiler/src/check.rs
use {super::Compiler, tangram_client::prelude::*};

#[derive(Debug, serde::Serialize)]
pub struct Request {
	pub modules: Vec<tg::module::Data>,
}

#[derive(Debug, serde::Deserialize)]
pub struct Response {
	pub diagnostics: Vec<tg::diagnostic::Data>,
}

impl Compiler {
	/// Get all diagnostics for the provided modules.
	pub async fn check(&self, modules: Vec<tg::module::Data>) -> tg::Result<Vec<tg::Diagnostic>> {
		// Create the request.
		let request = super::Request::Check(Request { modules });

		// Perform the request.
		let response = self.request(request).await?.unwrap_check();

		// Convert diagnostics from data to the non-serializable form.
		let diagnostics = response
			.diagnostics
			.into_iter()
			.map(TryInto::try_into)
			.collect::<tg::Result<Vec<_>>>()?;

		Ok(diagnostics)
	}
}
